### ISSUE TITLE
Deleting taxon with children deletes also taxons that aren't children of the deleted one

### ISSUE THREAD

**Issue by pamil** [Sylius/Sylius#3855]
It is not wise to have an entity that uses Softdeleteable and Tree at the one time, the whole tree structure gets messed up.

This happens when we try to remove taxonomy "Category" (with its root taxon).
```
PRE DELETE

mysql> select * from sylius_taxon;
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
| id  | taxonomy_id | parent_id | code | tree_left | tree_right | tree_level | deleted_at | path |
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
| 211 |          43 |      NULL | RTX1 |         1 |         18 |          0 | NULL       | NULL | <-- Category
| 212 |          44 |      NULL | RTX2 |        19 |         20 |          0 | NULL       | NULL | <-- Brand
| 213 |          43 |       211 | TX1  |         2 |         11 |          1 | NULL       | NULL |
| 214 |          43 |       213 | TX3  |         3 |          4 |          2 | NULL       | NULL |
| 215 |          43 |       213 | TX4  |         5 |         10 |          2 | NULL       | NULL |
| 216 |          43 |       215 | TX5  |         6 |          7 |          3 | NULL       | NULL |
| 217 |          43 |       215 | TX6  |         8 |          9 |          3 | NULL       | NULL |
| 218 |          43 |       211 | TX2  |        12 |         17 |          1 | NULL       | NULL |
| 219 |          43 |       218 | TX7  |        13 |         14 |          2 | NULL       | NULL |
| 220 |          43 |       218 | TX8  |        15 |         16 |          2 | NULL       | NULL |
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
10 rows in set (0.00 sec)

DELETE

mysql> select * from sylius_taxon;
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
| id  | taxonomy_id | parent_id | code | tree_left | tree_right | tree_level | deleted_at | path |
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
| 211 |          43 |      NULL | RTX1 |         1 |          6 |          0 | NULL       | NULL | <-- Category
| 212 |          44 |      NULL | RTX2 |         7 |          8 |          0 | NULL       | NULL | <-- Brand
| 213 |          43 |       211 | TX1  |         2 |          7 |          1 | NULL       | NULL |
| 214 |          43 |       213 | TX3  |         3 |          4 |          2 | NULL       | NULL |
| 215 |          43 |       213 | TX4  |         3 |          6 |          2 | NULL       | NULL |
| 216 |          43 |       215 | TX5  |         4 |          5 |          3 | NULL       | NULL |
| 217 |          43 |       215 | TX6  |         6 |          7 |          3 | NULL       | NULL |
| 218 |          43 |       211 | TX2  |         8 |          5 |          1 | NULL       | NULL |
| 219 |          43 |       218 | TX7  |         9 |          8 |          2 | NULL       | NULL |
| 220 |          43 |       218 | TX8  |         9 |         10 |          2 | NULL       | NULL |
+-----+-------------+-----------+------+-----------+------------+------------+------------+------+
10 rows in set (0.00 sec)

Query that gets children elements to be deleted:

SELECT node FROM Sylius\Component\Core\Model\Taxon node WHERE node.left BETWEEN ?1 AND ?2

SELECT s0_.code AS code0, s0_.tree_left AS tree_left1, s0_.tree_right AS tree_right2, s0_.tree_level AS tree_level3, s0_.deleted_at AS deleted_at4, s0_.id AS id5, s0_.path AS path6, s0_.taxonomy_id AS taxonomy_id7, s0_.parent_id AS parent_id8 FROM sylius_taxon s0_ WHERE (s0_.tree_left BETWEEN ? AND ?) AND (s0_.deleted_at IS NULL)

?1 = 2
?2 = 11

POST DELETE

mysql> select * from sylius_taxon;
+-----+-------------+-----------+------+-----------+------------+------------+---------------------+------+
| id  | taxonomy_id | parent_id | code | tree_left | tree_right | tree_level | deleted_at          | path |
+-----+-------------+-----------+------+-----------+------------+------------+---------------------+------+
| 211 |        NULL |      NULL | RTX1 |         1 |          6 |          0 | 2016-01-12 14:59:17 | NULL | <-- Category
| 212 |          44 |      NULL | RTX2 |         7 |          8 |          0 | 2016-01-12 14:59:17 | NULL | <-- Brand
| 213 |        NULL |       211 | TX1  |         2 |          7 |          1 | 2016-01-12 14:59:17 | NULL |
| 214 |        NULL |       213 | TX3  |         3 |          4 |          2 | 2016-01-12 14:59:17 | NULL |
| 215 |        NULL |       213 | TX4  |         3 |          6 |          2 | 2016-01-12 14:59:17 | NULL |
| 216 |        NULL |       215 | TX5  |         4 |          5 |          3 | 2016-01-12 14:59:17 | NULL |
| 217 |        NULL |       215 | TX6  |         6 |          7 |          3 | 2016-01-12 14:59:18 | NULL |
| 218 |        NULL |       211 | TX2  |         8 |          5 |          1 | 2016-01-12 14:59:18 | NULL |
| 219 |        NULL |       218 | TX7  |         9 |          8 |          2 | 2016-01-12 14:59:18 | NULL |
| 220 |        NULL |       218 | TX8  |         9 |         10 |          2 | 2016-01-12 14:59:18 | NULL |
+-----+-------------+-----------+------+-----------+------------+------------+---------------------+------+
10 rows in set (0.00 sec)
```

**Comment by michalmarcinkowski** [comment]
Will be fixed in #3773.

**Comment by michalmarcinkowski** [comment]
@pamil thanks for detailed report! :+1: Closing as it's a duplicate of #478.

**Comment by DmitriyTrt** [comment]
I got a similar issue on current Sylius v1.0.0-beta.1, it's even reproducible on a demo website.

Steps to reproduce on a demo website:
- Make a new root taxon.
- Add root child 1
- Add root child 2
- Add child of the child 1
- Delete child 1

Child 2 is also deleted. 

We also got a messed up tree on the test website, but I can't reproduce it yet. This is a quite critical issue, as it results in unexpected data loss. Not sure if I should submit a different issue.
